node/pkg/query/response.go

```diff
@@ -0,0 +1,507 @@
+
+	if err := ValidatePerChainQueryRequestType(queryType); err != nil {
+		return err
+	}
+
+	// Skip the response length.
+	var respLength uint32
+	if err := binary.Read(reader, binary.BigEndian, &respLength); err != nil {
+		return fmt.Errorf("failed to read response length: %w", err)
+	}
+
+	switch queryType {
+	case EthCallQueryRequestType:
+		r := EthCallQueryResponse{}
+		if err := r.UnmarshalFromReader(reader); err != nil {
+			return fmt.Errorf("failed to unmarshal eth call response: %w", err)
+		}
+		perChainResponse.Response = &r
+	default:
+		return fmt.Errorf("unsupported query type: %d", queryType)
+	}
+
+	return nil
+}
+
+// ValidatePerChainResponse performs basic validation on a per chain query response.
+func (perChainResponse *PerChainQueryResponse) Validate() error {
+	str := perChainResponse.ChainId.String()
+	if _, err := vaa.ChainIDFromString(str); err != nil {
+		return fmt.Errorf("invalid chainID: %d", uint16(perChainResponse.ChainId))
+	}
+
+	if perChainResponse.Response == nil {
+		return fmt.Errorf("response is nil")
+	}
+
+	if err := ValidatePerChainQueryRequestType(perChainResponse.Response.Type()); err != nil {
+		return err
+	}
+
+	if err := perChainResponse.Response.Validate(); err != nil {
+		return fmt.Errorf("chain specific response is invalid: %w", err)
+	}
+
+	return nil
+}
+
+// Equal checks for equality on two per chain query responses.
+func (left *PerChainQueryResponse) Equal(right *PerChainQueryResponse) bool {
+	if left.ChainId != right.ChainId {
+		return false
+	}
+
+	if left.Response == nil && right.Response == nil {
+		return true
+	}
+
+	if left.Response == nil || right.Response == nil {
+		return false
+	}
+
+	if left.Response.Type() != right.Response.Type() {
+		return false
+	}
+
+	switch leftEcq := left.Response.(type) {
+	case *EthCallQueryResponse:
+		switch rightEcd := right.Response.(type) {
+		case *EthCallQueryResponse:
+			return leftEcq.Equal(rightEcd)
+		default:
+			panic("unsupported query type on right") // We checked this above!
+		}
+	default:
+		panic("unsupported query type on left") // We checked this above!
+	}
+}
+
+//
+// Implementation of EthCallQueryResponse, which implements the ChainSpecificResponse for an EVM eth_call query response.
+//
+
+func (e *EthCallQueryResponse) Type() ChainSpecificQueryType {
+	return EthCallQueryRequestType
+}
+
+// Marshal serializes the binary representation of an EVM eth_call response.
+// This method calls Validate() and relies on it to range checks lengths, etc.
+func (ecr *EthCallQueryResponse) Marshal() ([]byte, error) {
+	if err := ecr.Validate(); err != nil {
+		return nil, err
+	}
+
+	buf := new(bytes.Buffer)
+	vaa.MustWrite(buf, binary.BigEndian, ecr.BlockNumber)
+	buf.Write(ecr.Hash[:])
+	vaa.MustWrite(buf, binary.BigEndian, ecr.Time.UnixMicro())
+
+	vaa.MustWrite(buf, binary.BigEndian, uint8(len(ecr.Results)))
+	for idx := range ecr.Results {
+		vaa.MustWrite(buf, binary.BigEndian, uint32(len(ecr.Results[idx])))
```

Comment:
ditto here

Reply:
Ditto with the response, it's checked in `Validate`.